# INTRODUCTION
This project analyzes a dataset containing information about flights in India, with the aim of conducting a statistical evaluation of flight characteristics to understand ticket prices according to said characteristics.

**Data source:**
kaggle [link](https://www.kaggle.com/datasets/shubhambathwal/flight-price-prediction), "Ease My Trip", Udemy  

## Tasks 
a) Does price vary with Airlines?  
b) How is the price affected when tickets are bought in just 1 or 2 days before departure?  
c) Does ticket price change based on the departure time and arrival time?  
d) How does the price change with change in Source and Destination?  
e) How does the ticket price vary between Economy and Business class?  

## Scope of the project
This study is part of a broader portfolio project designed to showcase my data analysis skills, including:
- Statistical reasoning
- Proficiency with core Python libraries (NumPy, Pandas, Matplotlib, Seaborn)
- Automation principles
- Overall analytical approach to working with structured data

Although the dataset was originally intended for building a predictive model, the main focus of this project is on:
- Data Cleaning
- Feature Engineering
- Exploratory Data Analysis (EDA)

These processes are used to extract meaningful insights and prepare the data for potential modeling, rather than building a predictive algorithm itself.

The entire study has been divided into three main sections (three deparate notebook files) to improve readability:

### Data Cleaning
This section demonstrates the process of cleaning a raw dataset by identifying inconsistencies, handling missing values, correcting formatting issues, and validating data integrity.  
In short, preparing the dataset for further analysis.

### Feature Engineering
Here, the dataset is transformed to include new or restructured variables, making it more suitable for analysis or modeling tasks. While not all engineered features are necessary for the final EDA, they are included to demonstrate relevant techniques and problem-solving capabilities

### Exploratory Data Analysis
This is where key tasks are addressed through statistical evaluation and visualizations. Various trends, distributions, and relationships are explored to extract insights and support data-driven reasoning.

Each task within these three main sections is documented with explanations of the relevant actions and insights.

**NOTE**: as the project is intended for showcasing and practicing purposes, some steps (mainly in the FE section) are not strictly necessary for the completion of the tasks or the final analysis.  
These intentional additions are meant to show familiraity with different data-preparation techniques.  
Additonally, while data cleaning and the feature engineering are often to be intertwined and occur simultenously in a real workflow, separating them into distinct sections was a deliberate choice to improve structure and clarity

# DATA CLEANING

In [7]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

In [543]:
df = pd.read_excel('flight_price.xlsx')

### 1. UNDERSTANDING THE DATA STRUCTURE

#### 1.1 General info

preliminary exploration of the dataset

In [547]:
# Displaying the dataset!
df

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [548]:
# Retrieving the total number of rows (records) and columns (features)
df.shape

(10683, 11)

In [549]:
# Retrieving general information about missing values and data dtype
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


#### Insights - general info:
The dataset contains 11 features and 10683 records.  
At this moment, the features (10) are mostly in the form of strings, including those containing dates and times or values that can be converted to numbers (Total_Stops): one of the main objective of the data cleaning process will be to convert such features into the correct data type.  
The only column whose values are stored as integers is 'Price', which is also the target variable.
The dataset is quite consistent in terms of missing values, with only two missing values: 1 in 'Route' and 1 in 'Total_Stops'.

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
#### 1.2 Finding unique values

Checking unique values and their count for categorical features

In [557]:
df['Airline'].nunique()

12

In [559]:
df['Airline'].unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [561]:
df['Source'].nunique()

5

In [563]:
df['Source'].unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

In [565]:
df['Destination'].nunique()

6

In [567]:
df['Destination'].unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

In [569]:
df['Route'].nunique() # There are too many different Route: in this case looking at the unique values is useless

128

In [571]:
df['Total_Stops'].nunique()

5

In [573]:
df['Total_Stops'].unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [575]:
df['Additional_Info'].nunique()

10

In [577]:
df['Additional_Info'].unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

In [579]:
# Checking unique values for numerical features

In [581]:
df['Date_of_Journey'].nunique()

44

In [583]:
df['Dep_Time'].nunique()

222

In [585]:
df['Arrival_Time'].nunique()

1343

In [587]:
df['Duration'].nunique()

368

In [589]:
df['Price'].nunique()

1870

#### Insights - unique values
In this dataset, many values in the numerical features — such as Price — are rarely repeated. As a result, applying .unique() on these columns would return a very long list of distinct values, which is not easily interpretable. In these cases, it is often more useful to use .nunique() to simply count the number of unique entries, which helps assess whether a .unique() check is worth running at all.

On the other hand, categorical features typically contain a limited set of unique values. For instance, the Total_Stops column likely includes only a few distinct entries, making it helpful to examine both the unique values and their count.
However, this assumption has limitations. Some categorical features — like Route — can still have a high number of unique values (e.g., 129 distinct routes), in which case listing them all is not particularly insightful.

### INSIGHTS - DATA STRUCTURE

Features:

1) **Airline**: categorical feature with the name of the airlines containing 12 unique values in the form of strings
   
2) **Date_of_Journey**: date feature with the date of departure in the form of string - to be converted to date

   
3) **Source**: categorical feature with the name of the source city containing 6 unique values in the form of string

 
4) **Destination**: categorical feature with the name of the destination city containing 6 unique values in the form of string


5) **Route**: categorical feature with the acronyms of the airports (source, stops and destination) in the form of string - 1 missing value


6) **Dep_Time**: date categorical feature with the time of departure in the form of string and format 'HH:MM' - to be converted to time


7) **Arrival_Time**: date feature with the time and (in some cases) date of arrival in the form of string and format 'HH:MM dd mm' - to be converted to time/datetime

 
8) **Duration**: shows the total duration of each flight (including stops)in the form of strings and format 'xx'h 'yy'm - to be converted to 'HH:MM'

 
9) **Total_Stops**: discrete numerical feature derived from 'Route' with the number of stops containing 5 unique values in the form of string - 1 missing value; to be converted to integer

 
10) **Additiona_Info**: categorical value with information about the flight (meal, layovers, baggages, etc) containing 10 unique values in the form of string

    
11) **Price**: numerical feature with the price of every flight. 

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
After inspecting the dataset, a copy of it will be cretated for manipulation purposes, so to leave the original one as is.

In [594]:
df_copy = df.copy()

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

### 2.  IDENTIFYING AND HANDLING DUPLICATE ROWS
Identifying and dropping duplicate rows helps eliminate noise when performing statistical analysis, as duplicates can distort metrics such as mean, mode, and standard deviation, and can skew the shape of distributions

#### 2.1 Finding duplicate rows

In [599]:
# Total number of rows (records) and Columns (features)
df_copy.shape

(10683, 11)

In [601]:
# Number of duplicated rows, excluding the first row
duplicated = df_copy.duplicated()  #default parameter for duplicated: keep = 'first'
duplicated.sum()

220

In [603]:
# Total number of duplicated rows, inclduing the first record
dups = df_copy.duplicated(keep = False) 
dups.sum()

415

In [605]:
# TEST: on one feature (Airline)
df_copy[df_copy.duplicated(keep=False)]['Airline'].sort_values(ascending = False)

2634      Vistara
5470      Vistara
1535     SpiceJet
2870     SpiceJet
7395     SpiceJet
          ...    
8640    Air India
1786    Air India
8613    Air India
8602    Air India
5483    Air India
Name: Airline, Length: 415, dtype: object

In [607]:
# Test: Ordering duplicated features by Airline
df_copy[df_copy.duplicated(keep=False)].sort_values(by= 'Airline', ascending = False)

Airline Date_of_Journey    Source Destination                  Route  \
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
1535   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
7395   SpiceJet      03/03/2019  Banglore   New Delhi              BLR → DEL   
...         ...             ...       ...         ...                    ...   
8640  Air India       3/06/2019     Delhi      Cochin  DEL → AMD → BOM → COK   
1786  Air India      15/06/2019     Delhi      Cochin  DEL → AMD → BOM → COK   
8613  Air India      27/06/2019     Delhi      Cochin  DEL → AMD → BOM → COK   
8602  Air India      27/06/2019     Delhi      Cochin  DEL → AMD → BOM → COK   
5483  Air India      27/05/2019     Delhi      Cochin  DEL → GOI → BOM → COK   

     Dep_Time  Arrival_Time Duration Total_Stops  \
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   
1535    20:30         23:20   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
7395    20:30         23:20   2h 50m    non-stop   
...       ...           ...      ...         ...   
8640    19:45  19:15 04 Jun  23h 30m     2 stops   
1786    19:45  19:15 16 Jun  23h 30m     2 stops   
8613    16:40  19:15 28 Jun  26h 35m     2 stops   
8602    19:45  19:15 28 Jun  23h 30m     2 stops   
5483    22:00  19:15 28 May  21h 15m     2 stops   

                   Additional_Info  Price  
2634                       No info   5403  
5470                       No info   5403  
1535  No check-in baggage included   3873  
2870  No check-in baggage included   4273  
7395                       No info   6860  
...                            ...    ...  
8640                       No info  10441  
1786                       No info   9653  
8613                       No info   9653  
8602                       No info   9653  
5483                       No info  10231  

[415 rows x 11 columns]

In [609]:
# Percentage of repeated records (excluding the first one)
(220/10683)*100

2.059346625479734

#### Insights - duplicate rows
Number of duplicated rows:
- 220 excluding the first/original record
- 415 including the first record

This indicates that some records have more than one duplicate.  

In this specific dataset, duplicates don't add any important information to the analysis as they are not a simptom of repeated behaviours: if a flight has the same identical info as another one (airline, date of departure and arrival, source, destination, etc.), they are necessarely the same flight as there can not be two identical flights at the same time.  

Additionally, the percentage of duplicated records excluding the first is 2.0% of the total, which seems a negligible amount.  

For these reasons, removal of duplicate shouldn't affect the study.

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
#### 2.2 Removing duplicate rows

In [613]:
# Dropping all the duplicates rows
df_copy = df_copy.drop_duplicates(keep = 'first')

In [615]:
# The following 2 are a check to make sure the process worked as intended
df_copy.shape

(10463, 11)

In [617]:
# Further check on elimination of duplicate rows
df_copy.duplicated().sum()

0

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
### 3. IDENTIFYING AND HANDLING MISSING VALUES (preliminary check)
Missing values have a specular effect compared to duplicates and the results of keeping them when not necessary or not handling them properly through statistical methods can also negatively affect statistical evaluations

df.info() has already indicated the presence of only two missing values in total: one in Route and one in Total_Stops.  
The next step will be about retrieving the related rows and evaluating how to handle them.

In [620]:
# Checking for missing values in Route
df_copy[df_copy['Route'].isnull()]

Airline Date_of_Journey Source Destination Route Dep_Time  \
9039  Air India       6/05/2019  Delhi      Cochin   NaN    09:45   

      Arrival_Time Duration Total_Stops Additional_Info  Price  
9039  09:25 07 May  23h 40m         NaN         No info   7480

In [622]:
# Checking for missing values in Total_Stops
df_copy[df_copy['Total_Stops'].isnull()]

Airline Date_of_Journey Source Destination Route Dep_Time  \
9039  Air India       6/05/2019  Delhi      Cochin   NaN    09:45   

      Arrival_Time Duration Total_Stops Additional_Info  Price  
9039  09:25 07 May  23h 40m         NaN         No info   7480

Both missing records for the two features appear to be on the same row

In [625]:
# Checking both columns with logical operator '|' (= 'or') to confirm NaN is on the same row 
df_copy[df_copy['Route'].isnull() | df_copy['Total_Stops'].isnull()]

Airline Date_of_Journey Source Destination Route Dep_Time  \
9039  Air India       6/05/2019  Delhi      Cochin   NaN    09:45   

      Arrival_Time Duration Total_Stops Additional_Info  Price  
9039  09:25 07 May  23h 40m         NaN         No info   7480

#### INSIGHTS - missing values
As stated, there are only two missing values in the dataset, one in route and one in destination.  
Additionally, the check performed on both features indicates that such values belong to the same row — which corresponds to a single record.  
All this suggests that dropping said row is a safe choice.

In fact:  
- Only one record is affected, so no significant data will be lost.

- Imputation would be unreliable here, since there's not enough context to guess them.

In [628]:
#Dropping the row with missing values: 9039 is the row with the missing value (nan)

df_copy.drop(9039, inplace = True) # df_copy = df_copy.drop(9039)  

In [630]:
# Checking if the row was correctly dropped
df_copy[df_copy['Route'].isna()]

Empty DataFrame
Columns: [Airline, Date_of_Journey, Source, Destination, Route, Dep_Time, Arrival_Time, Duration, Total_Stops, Additional_Info, Price]
Index: []

### 4. FIXING FORMAT AND STRUCTURAL ISSUES
This process aims to standardize data formats to ensure consistency in the dataset and enable proper comparisons and operations across different columns with similar features using vectorized methods.

#### 4.1 Extracting time from Arrival_Time

In [634]:
# Slippting minutes on the space and retrieving time from Arrival_Time
df_copy['Arrival_Time_Hour'] = df_copy['Arrival_Time'].str.split(' ').str[0]
df_copy['Arrival_Time_Hour']

0        01:10
1        13:15
2        04:25
3        23:30
4        21:35
         ...  
10678    22:25
10679    23:20
10680    11:20
10681    14:10
10682    19:15
Name: Arrival_Time_Hour, Length: 10462, dtype: object

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
#### 4.2 Extracting time from Duration

In [637]:
# Replacing symbols in Duration 
df_copy['Duration']=df['Duration'].str.replace('h ',':')
df_copy['Duration'] = df_copy['Duration'].str.replace('h', ':00')
df_copy['Duration'] = df_copy['Duration'].str.replace('m','')
df_copy['Duration']

0         2:50
1         7:25
2        19:00
3         5:25
4         4:45
         ...  
10678     2:30
10679     2:35
10680     3:00
10681     2:40
10682     8:20
Name: Duration, Length: 10462, dtype: object

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
#### 4.3 Extracting the number of stops from Total_Stops

In [640]:
# Replacing words and splitting/retrieving strings in Total_Stops
df_copy['Total_Stops'] = df_copy['Total_Stops'].str.replace('non-stop','0')
df_copy['Total_Stops'] = df_copy['Total_Stops'].str.split(' ').str[0]

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
### 5. CONVERTING DATA TYPES
The conversion completes the normalization process by converting data in the appropriate types, allowing to perform comparions and other operations across different columns with similar features using vectorized methods.

#### 5.1 Converting Date_of_Journey from string to timestamp

In [644]:
df_copy['Date_of_Journey'] = pd.to_datetime(df_copy['Date_of_Journey'])
df_copy['Date_of_Journey']

/var/folders/gs/d24257s15q1bzbn_vyvcxhb00000gp/T/ipykernel_96221/651070924.py:1: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df_copy['Date_of_Journey'] = pd.to_datetime(df_copy['Date_of_Journey'])


0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10462, dtype: datetime64[ns]

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
#### 5.2 Converting Dep_Time from string to datestamp (HH:MM)
Here and in the next step, 2 columns will be created:
- one with just the time (H:M) for clarity when reading the dataset
- one with a dummy date, which is necessary to perform vectorized operations between times in pandas

Once the Feature Ingenireeing process is performed, it will be easy to decide which one to keep

##### 5.2.1. Creating a column whit a dummy date

In [648]:
# Creating a new column with a dummy Date
df_copy['Dep_Time_Dummy']=pd.to_datetime(df_copy['Dep_Time'], format = '%H:%M')
df_copy[['Dep_Time_Dummy']]

Dep_Time_Dummy
0     1900-01-01 22:20:00
1     1900-01-01 05:50:00
2     1900-01-01 09:25:00
3     1900-01-01 18:05:00
4     1900-01-01 16:50:00
...                   ...
10678 1900-01-01 19:55:00
10679 1900-01-01 20:45:00
10680 1900-01-01 08:20:00
10681 1900-01-01 11:30:00
10682 1900-01-01 10:55:00

[10462 rows x 1 columns]

##### 5.2.2. Creating a column whit just the time

In [651]:
# Creating a column with just the time
df_copy['Dep_Time_Hour']=pd.to_datetime(df_copy['Dep_Time']).dt.time  # the parameter format = '%H:%M' is optional
df_copy[['Dep_Time_Hour']]

/var/folders/gs/d24257s15q1bzbn_vyvcxhb00000gp/T/ipykernel_96221/1711639493.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_copy['Dep_Time_Hour']=pd.to_datetime(df_copy['Dep_Time']).dt.time  # the parameter format = '%H:%M' is optional


Dep_Time_Hour
0          22:20:00
1          05:50:00
2          09:25:00
3          18:05:00
4          16:50:00
...             ...
10678      19:55:00
10679      20:45:00
10680      08:20:00
10681      11:30:00
10682      10:55:00

[10462 rows x 1 columns]

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
#### 5.3 Converting Arrival_Time from string to timestamp (HH:MM)
Since records in Arrival_Time present, along with the time, the date of arrival - for flights that have a shift in the day of arrival - the first step will be to separate these two variables

##### 5.3.1. Creating a column with dummy date

In [655]:
#column with the dummy date
df_copy['Arrival_Time_Dummy'] = pd.to_datetime(df_copy['Arrival_Time_Hour'], format = '%H:%M')
df_copy['Arrival_Time_Dummy']

0       1900-01-01 01:10:00
1       1900-01-01 13:15:00
2       1900-01-01 04:25:00
3       1900-01-01 23:30:00
4       1900-01-01 21:35:00
                ...        
10678   1900-01-01 22:25:00
10679   1900-01-01 23:20:00
10680   1900-01-01 11:20:00
10681   1900-01-01 14:10:00
10682   1900-01-01 19:15:00
Name: Arrival_Time_Dummy, Length: 10462, dtype: datetime64[ns]

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

##### 5.3.2 Crating a column whit just the time

In [659]:
#Column with just the time with .dt.time
df_copy['Arrival_Time_Hour'] = pd.to_datetime(df_copy['Arrival_Time_Hour'], format = '%H:%M').dt.time
df_copy['Arrival_Time_Hour'] 

0        01:10:00
1        13:15:00
2        04:25:00
3        23:30:00
4        21:35:00
           ...   
10678    22:25:00
10679    23:20:00
10680    11:20:00
10681    14:10:00
10682    19:15:00
Name: Arrival_Time_Hour, Length: 10462, dtype: object

In [661]:
df_copy.head(5)

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      2019-03-24  Banglore   New Delhi              BLR → DEL   
1    Air India      2019-05-01   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways      2019-06-09     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      2019-05-12   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      2019-03-01  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  \
0    22:20  01:10 22 Mar     2:50           0         No info   3897   
1    05:50         13:15     7:25           2         No info   7662   
2    09:25  04:25 10 Jun    19:00           2         No info  13882   
3    18:05         23:30     5:25           1         No info   6218   
4    16:50         21:35     4:45           1         No info  13302   

  Arrival_Time_Hour      Dep_Time_Dummy Dep_Time_Hour  Arrival_Time_Dummy  
0          01:10:00 1900-01-01 22:20:00      22:20:00 1900-01-01 01:10:00  
1          13:15:00 1900-01-01 05:50:00      05:50:00 1900-01-01 13:15:00  
2          04:25:00 1900-01-01 09:25:00      09:25:00 1900-01-01 04:25:00  
3          23:30:00 1900-01-01 18:05:00      18:05:00 1900-01-01 23:30:00  
4          21:35:00 1900-01-01 16:50:00      16:50:00 1900-01-01 21:35:00

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
#### 5.4 Converting Total_Stops from string to integer

In [664]:
df_copy['Total_Stops'] = df_copy['Total_Stops'].astype(int)

In [666]:
df_copy['Total_Stops'].unique()

array([0, 2, 1, 3, 4])

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

#### 5.5 Column name changes
This step is mainly performed to improve readability and make column names easier to reference in the following steps

In [670]:
df_copy.rename(columns = {
    'Date_of_Journey': 'Date of Journey', 
    'Dep_Time':'Dep Time', 
    'Arrival_Time':'Arr Time', 
    'Total_Stops': 'Total Stops',
    'Additional_Info':'Additional Info',
}, inplace = True)

In [672]:
df_copy.head(2)

Airline Date of Journey    Source Destination                  Route  \
0     IndiGo      2019-03-24  Banglore   New Delhi              BLR → DEL   
1  Air India      2019-05-01   Kolkata    Banglore  CCU → IXR → BBI → BLR   

  Dep Time      Arr Time Duration  Total Stops Additional Info  Price  \
0    22:20  01:10 22 Mar     2:50            0         No info   3897   
1    05:50         13:15     7:25            2         No info   7662   

  Arrival_Time_Hour      Dep_Time_Dummy Dep_Time_Hour  Arrival_Time_Dummy  
0          01:10:00 1900-01-01 22:20:00      22:20:00 1900-01-01 01:10:00  
1          13:15:00 1900-01-01 05:50:00      05:50:00 1900-01-01 13:15:00

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

### 6 CONCISTENCY CHECK ACROSS COLUMNS
It's important to make sure that there are no inconsistent values in the features, otherwise operations and modelling couldn't work properly

#### 6.1 Consistency check on Price column 

In [677]:
# Checking for negative prices
df_copy[df_copy['Price']<0]

Empty DataFrame
Columns: [Airline, Date of Journey, Source, Destination, Route, Dep Time, Arr Time, Duration, Total Stops, Additional Info, Price, Arrival_Time_Hour, Dep_Time_Dummy, Dep_Time_Hour, Arrival_Time_Dummy]
Index: []

In [679]:
df_copy[['Price']].describe()

Price
count  10462.000000
mean    9026.790289
std     4624.849541
min     1759.000000
25%     5224.000000
50%     8266.000000
75%    12344.750000
max    79512.000000

In [681]:
df_copy[(df_copy['Price']>=50000) & (df_copy['Price']<=79512)].shape

(8, 15)

In [683]:
df_copy.shape

(10462, 15)

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

In [686]:
#Checking if all values in Price are integers 
(df_copy['Price'] % 1 == 0)

0        True
1        True
2        True
3        True
4        True
         ... 
10678    True
10679    True
10680    True
10681    True
10682    True
Name: Price, Length: 10462, dtype: bool

In [688]:
(df_copy['Price'] % 1 == 0).shape

(10462,)

In [690]:
(df_copy['Price'] % 1 == 0).all()

True

In [692]:
df_copy['Price'].dtype

dtype('int64')

In [694]:
df_copy['Price'].dtype == 'float64'

#or, if we wanted to check the amount of records with dtype = float

# (df_copy['Price'] % 1 != 0).all()
# (df_copy['Price'] % 1 != 0).shape

False

#### Insights
The records in Price are all positive!

NOTE: The last checks were reduntant as we know the dtype from df_copy.info().  
If there were a mix of int and float, df.info() would return float, in which case it could be useful to check the number of float and integers

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

#### 6.2 Consistency check on Source and Destination

In [700]:
df_copy[df_copy['Source']==df_copy['Destination']]

Empty DataFrame
Columns: [Airline, Date of Journey, Source, Destination, Route, Dep Time, Arr Time, Duration, Total Stops, Additional Info, Price, Arrival_Time_Hour, Dep_Time_Dummy, Dep_Time_Hour, Arrival_Time_Dummy]
Index: []

#### Insights
No inconcistencies were found between Source and Destination columns as there are no identical destination for each source and viceversa!

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
#### 6.3 Consistency check on Duration

#### 6.3.1 Checking possible negative Durations

In [705]:
df_copy[(df_copy['Duration'].str.replace(':','').astype(int))<0]

#or, I could have created a new column and ran the check on it (less efficient)

# df_copy['Duration']=df_copy['Duration'].str.replace(':','').astype(int)
# df_copy[df_copy['Duration']<0]

Empty DataFrame
Columns: [Airline, Date of Journey, Source, Destination, Route, Dep Time, Arr Time, Duration, Total Stops, Additional Info, Price, Arrival_Time_Hour, Dep_Time_Dummy, Dep_Time_Hour, Arrival_Time_Dummy]
Index: []

#### Insights
Records in duration are all positive!

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

#### 6.3.2 Checking for extremes durations

In [710]:
pd.to_datetime(df_copy['Duration']).min()

/var/folders/gs/d24257s15q1bzbn_vyvcxhb00000gp/T/ipykernel_96221/3095469934.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(df_copy['Duration']).min()


OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 21:5, at position 7

The previous check has immediately highlitghted an inconcistency in the time format of Duration where 21h 5m has been converted to 21:5 which is not a proper format for time. There might be more entries that have been converted to such format. This kind of problems can be easily fixed durin the FE process by separating the plitting the column into hours and minutes. 
For the sake completeness we'll address them in this section

In [713]:
#Check for the number of record with strings that contain only 4 digits (2 numbers then : and another number)
df_copy[df_copy['Duration'].str.split(':').str[1].str.len()==1] #.shape

Airline Date of Journey    Source Destination  \
7            Jet Airways      2019-03-01  Banglore   New Delhi   
40     Multiple carriers      2019-05-21     Delhi      Cochin   
42           Jet Airways      2019-05-18     Delhi      Cochin   
48             Air India      2019-05-15   Kolkata    Banglore   
56             Air India      2019-03-01  Banglore   New Delhi   
...                  ...             ...       ...         ...   
10592        Jet Airways      2019-05-21   Kolkata    Banglore   
10613        Jet Airways      2019-06-03  Banglore       Delhi   
10629  Multiple carriers      2019-03-03     Delhi      Cochin   
10637        Jet Airways      2019-04-24  Banglore       Delhi   
10663        Jet Airways      2019-06-06     Delhi      Cochin   

                       Route Dep Time      Arr Time Duration  Total Stops  \
7            BLR → BOM → DEL    08:00  05:05 02 Mar     21:5            1   
40           DEL → HYD → COK    07:05         18:10     11:5            1   
42           DEL → BOM → COK    20:55  19:00 19 May     22:5            1   
48           CCU → HYD → BLR    19:00  11:05 16 May     16:5            1   
56     BLR → BOM → AMD → DEL    08:50         23:55     15:5            2   
...                      ...      ...           ...      ...          ...   
10592        CCU → BOM → BLR    18:55  12:00 22 May     17:5            1   
10613              BLR → DEL    18:55         22:00      3:5            0   
10629        DEL → HYD → COK    07:05         16:10      9:5            1   
10637              BLR → DEL    18:55         22:00      3:5            0   
10663  DEL → JAI → BOM → COK    19:30  12:35 07 Jun     17:5            2   

                   Additional Info  Price Arrival_Time_Hour  \
7                          No info  22270          05:05:00   
40                         No info   9646          18:10:00   
42     In-flight meal not included  12373          19:00:00   
48                         No info   6117          11:05:00   
56                         No info  17345          23:55:00   
...                            ...    ...               ...   
10592  In-flight meal not included  10844          12:00:00   
10613                      No info   8016          22:00:00   
10629                      No info   7905          16:10:00   
10637                      No info   7229          22:00:00   
10663  In-flight meal not included  11733          12:35:00   

           Dep_Time_Dummy Dep_Time_Hour  Arrival_Time_Dummy  
7     1900-01-01 08:00:00      08:00:00 1900-01-01 05:05:00  
40    1900-01-01 07:05:00      07:05:00 1900-01-01 18:10:00  
42    1900-01-01 20:55:00      20:55:00 1900-01-01 19:00:00  
48    1900-01-01 19:00:00      19:00:00 1900-01-01 11:05:00  
56    1900-01-01 08:50:00      08:50:00 1900-01-01 23:55:00  
...                   ...           ...                 ...  
10592 1900-01-01 18:55:00      18:55:00 1900-01-01 12:00:00  
10613 1900-01-01 18:55:00      18:55:00 1900-01-01 22:00:00  
10629 1900-01-01 07:05:00      07:05:00 1900-01-01 16:10:00  
10637 1900-01-01 18:55:00      18:55:00 1900-01-01 22:00:00  
10663 1900-01-01 19:30:00      19:30:00 1900-01-01 12:35:00  

[619 rows x 15 columns]

We have 619 record that match the format hh:m

In order to check if Duration is Consistent with the numbers of stops we must convert such records into hh:mm, making sure that the first minute is m=0

In [716]:
df_copy['Duration'] = df_copy['Duration'].apply(lambda x: x if ':' not in x else x.split(':')[0] + ':' + x.split(':')[1].zfill(2))


#or (my version)

#df_copy['Duration'].apply(lambda x: x if ':' not in x else x.split(':')[0]+ ':0'+ x.split(':')[1] if len(x.split(':')[1])==1 else x)

NOTE: .zfill(2) pads (=adds extra character) the string is referring to with a 0 (always on the left) if it’s less than 2 digits → '5' → '05'

In [719]:
df_copy.head(8)

Airline Date of Journey    Source Destination                  Route  \
0       IndiGo      2019-03-24  Banglore   New Delhi              BLR → DEL   
1    Air India      2019-05-01   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways      2019-06-09     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      2019-05-12   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      2019-03-01  Banglore   New Delhi        BLR → NAG → DEL   
5     SpiceJet      2019-06-24   Kolkata    Banglore              CCU → BLR   
6  Jet Airways      2019-03-12  Banglore   New Delhi        BLR → BOM → DEL   
7  Jet Airways      2019-03-01  Banglore   New Delhi        BLR → BOM → DEL   

  Dep Time      Arr Time Duration  Total Stops              Additional Info  \
0    22:20  01:10 22 Mar     2:50            0                      No info   
1    05:50         13:15     7:25            2                      No info   
2    09:25  04:25 10 Jun    19:00            2                      No info   
3    18:05         23:30     5:25            1                      No info   
4    16:50         21:35     4:45            1                      No info   
5    09:00         11:25     2:25            0                      No info   
6    18:55  10:25 13 Mar    15:30            1  In-flight meal not included   
7    08:00  05:05 02 Mar    21:05            1                      No info   

   Price Arrival_Time_Hour      Dep_Time_Dummy Dep_Time_Hour  \
0   3897          01:10:00 1900-01-01 22:20:00      22:20:00   
1   7662          13:15:00 1900-01-01 05:50:00      05:50:00   
2  13882          04:25:00 1900-01-01 09:25:00      09:25:00   
3   6218          23:30:00 1900-01-01 18:05:00      18:05:00   
4  13302          21:35:00 1900-01-01 16:50:00      16:50:00   
5   3873          11:25:00 1900-01-01 09:00:00      09:00:00   
6  11087          10:25:00 1900-01-01 18:55:00      18:55:00   
7  22270          05:05:00 1900-01-01 08:00:00      08:00:00   

   Arrival_Time_Dummy  
0 1900-01-01 01:10:00  
1 1900-01-01 13:15:00  
2 1900-01-01 04:25:00  
3 1900-01-01 23:30:00  
4 1900-01-01 21:35:00  
5 1900-01-01 11:25:00  
6 1900-01-01 10:25:00  
7 1900-01-01 05:05:00

Now we can turn the record in the Duration column into float so to perform comparison among them, including finding the min and the max

In [722]:
df_copy['Duration_todrop']=df_copy['Duration'].str.replace(':','.').astype(float)

In [724]:
df_copy['Duration_todrop'].min()

1.15

In [726]:
df_copy['Duration_todrop'].max()

47.4

In [728]:
df_copy[['Duration_todrop']].describe()

Duration_todrop
count     10462.000000
mean         10.308765
std           8.356638
min           1.150000
25%           2.500000
50%           8.250000
75%          15.100000
max          47.400000

In [730]:
# Quick check for extreme durations 
df_copy['Duration_todrop'].value_counts(ascending = False).sort_index(ascending = False).head(10).to_frame()

count
Duration_todrop       
47.40                1
47.00                1
42.45                1
42.05                1
41.20                1
40.20                1
39.05                1
38.35                4
38.20                4
38.15               12

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

#### NOTE - important Error:
From a previous analysis emerged a record for a flight with a duration of 5 minutes.  
During the convertion of Duration from the orignal format ('xx'H 'yy'H), all the entries with no hour have been changed into hours. This happened because the convertion didn't take into consideration datapoints lacking the hour. Though this is not a major issues since a 5 minutes flight is certainly impossible, it's conceptually wrong and could lead to inaccurate data since 5 minutes turned into 5 hours.

We are now going to run a second datacleaning for the Duration column from scratch

In [734]:
df_copy[df_copy['Duration']=='5m']

Empty DataFrame
Columns: [Airline, Date of Journey, Source, Destination, Route, Dep Time, Arr Time, Duration, Total Stops, Additional Info, Price, Arrival_Time_Hour, Dep_Time_Dummy, Dep_Time_Hour, Arrival_Time_Dummy, Duration_todrop]
Index: []

In [736]:
df_copy.loc[[6474]]

Airline Date of Journey  Source Destination                  Route  \
6474  Air India      2019-03-06  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep Time Arr Time Duration  Total Stops Additional Info  Price  \
6474    16:50    16:55        5            2         No info  17327   

     Arrival_Time_Hour      Dep_Time_Dummy Dep_Time_Hour  Arrival_Time_Dummy  \
6474          16:55:00 1900-01-01 16:50:00      16:50:00 1900-01-01 16:55:00   

      Duration_todrop  
6474              5.0

In [738]:
df_copy['Duration'].unique()

array(['2:50', '7:25', '19:00', '5:25', '4:45', '2:25', '15:30', '21:05',
       '25:30', '7:50', '13:15', '2:35', '2:15', '12:10', '26:35', '4:30',
       '22:35', '23:00', '20:35', '5:10', '15:20', '2:55', '13:20',
       '15:10', '5:45', '5:55', '13:25', '22:00', '5:30', '10:25', '5:15',
       '2:30', '6:15', '11:55', '11:05', '8:30', '22:05', '2:45', '12:00',
       '16:05', '19:55', '3:15', '25:20', '3:00', '16:15', '15:05',
       '6:30', '25:05', '12:25', '27:20', '10:15', '10:30', '1:30',
       '1:25', '26:30', '7:20', '13:30', '5:00', '19:05', '14:50', '2:40',
       '22:10', '9:35', '10:00', '21:20', '18:45', '12:20', '18:00',
       '9:15', '17:30', '16:35', '12:15', '7:30', '24:00', '8:55', '7:10',
       '14:30', '30:20', '15:00', '12:45', '10:10', '15:25', '14:05',
       '20:15', '23:10', '18:10', '16:00', '2:20', '8:00', '16:55',
       '3:10', '14:00', '23:50', '21:40', '21:15', '10:50', '8:15',
       '8:35', '11:50', '27:35', '8:25', '20:55', '4:50', '8:10', '24:25

The previous output doesn't really help us to immediately spot the 5m entry.  
We must find another way to check those records that are in a different format than the most common one

In [741]:
# Check for Duration with entries/strings that present a lenght smaller than the common format in the original dataset
df[df['Duration'].str.len()<3]

Airline Date_of_Journey    Source Destination  \
53                IndiGo      18/06/2019  Banglore       Delhi   
74                IndiGo      21/03/2019     Delhi      Cochin   
97           Jet Airways       3/06/2019  Banglore       Delhi   
130            Air India      21/03/2019  Banglore   New Delhi   
140    Multiple carriers       9/04/2019     Delhi      Cochin   
...                  ...             ...       ...         ...   
10496        Jet Airways       1/05/2019  Banglore       Delhi   
10512        Jet Airways      12/06/2019  Banglore       Delhi   
10529        Jet Airways      12/04/2019  Banglore       Delhi   
10530          Air India      03/03/2019  Banglore   New Delhi   
10680        Jet Airways      27/04/2019  Banglore       Delhi   

                 Route Dep_Time  Arrival_Time Duration Total_Stops  \
53           BLR → DEL    21:15  00:15 19 Jun       3h    non-stop   
74     DEL → BLR → COK    05:05         10:05       5h      1 stop   
97           BLR → DEL    19:50         22:50       3h    non-stop   
130    BLR → MAA → DEL    11:50         19:50       8h      1 stop   
140    DEL → BOM → COK    17:30  01:30 10 Apr       8h      1 stop   
...                ...      ...           ...      ...         ...   
10496        BLR → DEL    07:10         10:10       3h    non-stop   
10512        BLR → DEL    07:10         10:10       3h    non-stop   
10529        BLR → DEL    07:10         10:10       3h    non-stop   
10530  BLR → HYD → DEL    08:15         12:15       4h      1 stop   
10680        BLR → DEL    08:20         11:20       3h    non-stop   

                   Additional_Info  Price  
53                         No info   3943  
74                         No info   6893  
97     In-flight meal not included   6478  
130                        No info   5932  
140                        No info  13017  
...                            ...    ...  
10496  In-flight meal not included   4030  
10512  In-flight meal not included   5769  
10529                      No info   7229  
10530                      No info   6273  
10680                      No info   7229  

[516 rows x 11 columns]

The previous approach doesn't help either since also entries with only the hour (19h) has the same lenght as the ones with only minuts (5m)

In [744]:
# Checking which rows contain only minutes and not hours in the original dataset
df[(df['Duration'].str.contains('m')) & (~df['Duration'].str.contains('h'))]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

In [746]:
df[(df['Duration'].str.contains('m')) & (~df['Duration'].str.contains('h'))].shape


(1, 11)

This check found only one entry with this kind of Duration. It's reasonable to assume this was a standalone mistake and directly dropping will have no further consequences as any imputation method or analysis would be suprefluos: an imputation method for just one record could be potentially misleading!

#### NOTE: Accidents Happen
This incorrect record was found in a previous analysis by accident: in fact, while trying to change the data type of Duration, Python raised an error.
This is a clear sign that it's impossible to check every record perfectly and sometimes luck is the best help we can hope for

In [749]:
#Check the corresponding df row in df_copy
df_copy.loc[[6474]]

Airline Date of Journey  Source Destination                  Route  \
6474  Air India      2019-03-06  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep Time Arr Time Duration  Total Stops Additional Info  Price  \
6474    16:50    16:55        5            2         No info  17327   

     Arrival_Time_Hour      Dep_Time_Dummy Dep_Time_Hour  Arrival_Time_Dummy  \
6474          16:55:00 1900-01-01 16:50:00      16:50:00 1900-01-01 16:55:00   

      Duration_todrop  
6474              5.0

In [751]:
df_copy = df_copy.drop(6474)

#or

#df_copy.drop(6474, inplace = True)

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

Though we already found the minimum as a bad record, we can finally compare similar Durations to see if there's any duration that is too short compared to other similar flights.  

In [755]:
df_copy['Duration_todrop'].min()

1.15

In [757]:
# Comparing the minimum with the mode of the duration for the same flight
df_copy[(df_copy['Source']=='Mumbai') & (df_copy['Destination']=='Hyderabad') & (df_copy['Airline']=='Air India')]['Duration'].mode()

0    1:20
Name: Duration, dtype: object

#### Insights
Considering that the mode duration for flights from Mumbai to Hyderabad is 1 hour and 20 minutes, a minimum duration of 1 hour and 15 minutes appears plausible.

On the higher end of the spectrum, a maximum flight duration of 47.4 hours seems unusually long. While this could represent a rare but real occurrence (e.g., major delays or reroutes), further exploratory data analysis (EDA) will help determine whether it’s appropriate to treat this as an outlier and potentially exclude it from the dataset

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

#### 6.4 Consistency check between Route and Total Stops 

In [762]:
# Comparing Total Stops with number of airport (=stops) in Route
df_copy[df_copy['Total Stops']!=(df_copy['Route'].str.count('→')-1)].shape

(0, 16)

In [766]:
df_copy.head(3)

Airline Date of Journey    Source Destination                  Route  \
0       IndiGo      2019-03-24  Banglore   New Delhi              BLR → DEL   
1    Air India      2019-05-01   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways      2019-06-09     Delhi      Cochin  DEL → LKO → BOM → COK   

  Dep Time      Arr Time Duration  Total Stops Additional Info  Price  \
0    22:20  01:10 22 Mar     2:50            0         No info   3897   
1    05:50         13:15     7:25            2         No info   7662   
2    09:25  04:25 10 Jun    19:00            2         No info  13882   

  Arrival_Time_Hour      Dep_Time_Dummy Dep_Time_Hour  Arrival_Time_Dummy  \
0          01:10:00 1900-01-01 22:20:00      22:20:00 1900-01-01 01:10:00   
1          13:15:00 1900-01-01 05:50:00      05:50:00 1900-01-01 13:15:00   
2          04:25:00 1900-01-01 09:25:00      09:25:00 1900-01-01 04:25:00   

   Duration_todrop  
0             2.50  
1             7.25  
2            19.00

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

### Exporting Dataset so it keeps the data types implemented in this section

In [ ]:
df_copy.to_parquet('cleaned dataset.parquet')

### Attempt of automation of the data cleaning process (work in progress)

In [811]:
def data_cleaning(df, verbose = True):
    print('1.1 RETRIEVING GENERAL INFO ABOUT THE DATASET\n')
    df.info()
    print('\n-------------------------------------')
    print(f"\n1.2 TOTAL NUMBER OF ROWS AND COLUMN: {df.shape}")
    print('\n-------------------------------------')
    print('\n1.3 ASSESSING UNIQUE VALUES FOR EACH FEATURE')
    max_unique = int(input('What is the acceptable max number of unique values? '))
    for col in df:
        unique_count = df[col].nunique()
        print()
        print(f"Unique records count for '{col}': {unique_count}")
        
        if df[col].nunique() < max_unique:
            unique_value = df[col].unique()
            print(f"The unique values for {col} are \n{unique_value}")
            
        else:
            print(f"The number of unique records for {col} is too high to be easily interpretable")
            pass
    
    print ('\n-------------------------------------')
    print('\n CREATION OF A COPY OF THE DATASET TO WORK WITH')
    df_copy = df.copy()
    print ('\n-------------------------------------')
    
    print('\n2. IDENTIFYING AND HANDLING DUPLICATE ROWS\n')
    duplicate = df_copy.duplicated()
    print(f" Number of duplicate rows excluding the first one: {duplicate.sum()}")
    
    print('\nGiven the nature of this dataset - identical flights are not possible -'
          "it's convenient to drop duplicates") 
    df_copy = df_copy.drop_duplicates(keep = 'first')
    print(f"\nChecking the number of duplicates after elimination: {df_copy.duplicated().sum()}")
    print ('\n-------------------------------------')

    print('\n3. IDENTIFYING AND HANDLING MISSING RECORDS (preliminary check)')
    col_list = []
    for col in df_copy:
        if df_copy[col].isnull().any():
            col_list.append(col)
            missing_value = df_copy[col].isnull().sum()
            print(f"\n'{col}' has : {missing_value}")
        else:
            pass
    nan_rows = df_copy[df_copy['Route'].isnull() | df_copy['Total_Stops'].isnull()]
    print('\nAddressing the rows with missing values:\n') 
    print(nan_rows.to_string(index=False))

    print("\nBoth missing values are in the same row, dropping only row won't affect the analysis!")

    print ('\n-------------------------------------')

    print('\n4. FIXING FORMAT AND STRUCTURAL ISSUES')
    print('\n4.1 Extracting time from Arrival_Time')
    arrival_time_hour = df_copy['Arrival_Time_Hour'] = df_copy['Arrival_Time'].str.split(' ').str[0]
    display(arrival_time_hour)
    print('\n4.1 Extracting time from Duration')
    df_copy['Duration']=df['Duration'].str.replace('h ',':')
    df_copy['Duration'] = df_copy['Duration'].str.replace('h', ':00')
    duration = df_copy['Duration'] = df_copy['Duration'].str.replace('m','')
    display(duration)
    print('\n4.1 Extracting the number of stops from Total_Stops')
    df_copy['Total_Stops'] = df_copy['Total_Stops'].str.replace('non-stop','0')
    df_copy['Total_Stops'] = df_copy['Total_Stops'].str.split(' ').str[0]

In [813]:
data_cleaning(df)

1.1 RETRIEVING GENERAL INFO ABOUT THE DATASET

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB

-------------------------------------

1.2 TOTAL NUMBER OF ROWS AND COLUMN: (10683, 11)

-------------------------------------

1.3 ASSESSING UNIQUE VALUES FOR EACH FEATURE


What is the acceptable max number of unique values?  10



Unique records count for 'Airline': 12
The number of unique records for Airline is too high to be easily interpretable

Unique records count for 'Date_of_Journey': 44
The number of unique records for Date_of_Journey is too high to be easily interpretable

Unique records count for 'Source': 5
The unique values for Source are 
['Banglore' 'Kolkata' 'Delhi' 'Chennai' 'Mumbai']

Unique records count for 'Destination': 6
The unique values for Destination are 
['New Delhi' 'Banglore' 'Cochin' 'Kolkata' 'Delhi' 'Hyderabad']

Unique records count for 'Route': 128
The number of unique records for Route is too high to be easily interpretable

Unique records count for 'Dep_Time': 222
The number of unique records for Dep_Time is too high to be easily interpretable

Unique records count for 'Arrival_Time': 1343
The number of unique records for Arrival_Time is too high to be easily interpretable

Unique records count for 'Duration': 368
The number of unique records for Duration is too high to be ea

0        01:10
1        13:15
2        04:25
3        23:30
4        21:35
         ...  
10678    22:25
10679    23:20
10680    11:20
10681    14:10
10682    19:15
Name: Arrival_Time, Length: 10463, dtype: object


4.1 Extracting time from Duration


0         2:50
1         7:25
2        19:00
3         5:25
4         4:45
         ...  
10678     2:30
10679     2:35
10680     3:00
10681     2:40
10682     8:20
Name: Duration, Length: 10463, dtype: object


4.1 Extracting the number of stops from Total_Stops
